### PR TITLE
Add automatic downloading of extra containers (harvester/QA/reports support)

### PR DESCRIPTION
**Pull request by JackMc**
This also adds a redis container. It is not yet linked into the web container.

### CODE REVIEW

**Comment by JackMc** [comment]
This addresses (perhaps only partly) #52

**Comment by JackMc** [comment]
I would love any feedback on this

File: datacats/cli/manage.py [hunk 1]
```diff
@@ -240,10 +245,11 @@ def tweak(environment, opts):
     """Commands operating on environment data
 
 Usage:
-  datacats tweak --install-postgis [ENVIRONMENT]
+  datacats tweak [--add-redis|--install-postgis] [ENVIRONMENT]
```
Comment (wardi) [inline datacats/cli/manage.py]: not quite right, tweak with no options doesn't make sense

```
Usage:
  datacats tweak --install-postgis [ENVIRONMENT]
  datacats tweak --add-redis [ENVIRONMENT]
```

File: datacats/environment.py [hunk 2]
```diff
@@ -247,7 +252,7 @@ def start_supporting_containers(self):
         operate) if they aren't already running.
         """
         task.start_supporting_containers(self.sitedir, self.target,
-            self.passwords, self._get_container_name)
+            self.passwords, self.extra_containers, self._get_container_name)
```
Comment (wardi) [inline datacats/environment.py]: add new parameter to the end, please

File: datacats/task.py [hunk 3]
```diff
@@ -434,11 +439,16 @@ def create_source(srcdir, preload_image, datapusher=False):
             )
 
 
+# Maps container extra names to actual names
+EXTRA_IMAGE_MAPPING = {'redis': 'redis'}
+
+
 def start_supporting_containers(sitedir, srcdir, passwords,
-        get_container_name):
+        extra_containers, get_container_name):
```
Comment (wardi) [inline datacats/task.py]: er, I meant here

**Comment by JackMc** [comment]
In other news, the Harvester "New" page has a problem... A required field just doesn't exist:

**Comment by wardi** [comment]
is this the right version of the harvester?  If you're stuck on this, the other extensions are a bigger priority

**Comment by JackMc** [comment]
I was running master - do you think that might be the problem?

**Comment by JackMc** [comment]
I tried QA and it was giving me trouble only by the fact that we don't have Celery set up. It seems CKAN has some celery support built into its paster plugin, but it just seems to fail on me:

```
(ckan)shell@testenv:~$ paster --plugin=ckan celeryd
Traceback (most recent call last):
  File "/usr/lib/ckan/bin/paster", line 9, in <module>
    load_entry_point('PasteScript==1.7.5', 'console_scripts', 'paster')()
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/script/command.py", line 104, in run
    invoke(command, command_name, options, args[1:])
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/script/command.py", line 143, in invoke
    exit_code = runner.run(args)
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/script/command.py", line 238, in run
    result = self.command()
  File "/project/ckan/ckan/lib/cli.py", line 1000, in command
    self.run_()
  File "/project/ckan/ckan/lib/cli.py", line 1014, in run_
    os.environ['CKAN_CONFIG'] = os.path.abspath(self.options.config)
  File "/usr/lib/ckan/lib/python2.7/posixpath.py", line 367, in abspath
    if not isabs(path):
  File "/usr/lib/ckan/lib/python2.7/posixpath.py", line 61, in isabs
    return s.startswith('/')
AttributeError: 'NoneType' object has no attribute 'startswith'
```

**Comment by JackMc** [comment]
Same issue after checking out the stable branch

**Comment by deniszgonjanin** [comment]
Harvester is a generic plugin that also needs a specific harvester plugin for each 'source' to harvest from. It comes with a `ckan_harvester` plugin included. It needs to be enabled in the plugins:

`ckan.plugins = harvest ckan_harvester`

**Comment by wardi** [comment]
@JackMc this is good to be reviewed?

**Comment by JackMc** [comment]
Yes please

File: datacats/environment.py [hunk 4]
```diff
@@ -585,6 +593,26 @@ def fully_running(self):
...
+        """
+        Add a container as a 'extra'. These are running containers which are not necessary for
+        running default CKAN but are useful for certain extensions
+        :param container: The container name to add
+        :param error_on_exists: Raise a DatacatsError if the extra container already exists.
+        """
+        if container in self.extra_containers and error_on_exists:
+            raise DatacatsError('{} is already added as an extra container.'.format(container))
+
+        self.extra_containers.append(container)
```
Comment (wardi) [inline datacats/environment.py]: missing an `if container in self.extra_containers: return`

**Comment by wardi** [comment]
Do harvester, QA and reports work after this change, and after the user runs the `datacats tweak --add-redis` command?

**Comment by JackMc** [comment]
As long as you do the qa pull (or whatever their command is), harvester harvest and reports run command manually, yeah

**Comment by wardi** [comment]
the last thing I ask for is that this tweak command has a test added to see that it does its thing properly

**Comment by JackMc** [comment]
Like check that the container exists?

**Comment by wardi** [comment]
Check that the command completes and that on restart the container is running, yes

**Comment by deniszgonjanin** [comment]
I've been testing the redis container this morning, works great @JackMc :+1:

**Comment by JackMc** [comment]
Okie, gonna add a test

**Comment by JackMc** [comment]
Test added and passed!

**Comment by JackMc** [comment]
:smiley:

**Comment by florianm** [comment]
I just got stuck after `datacats tweak --add-redis` because I forgot I have to `datacats reload`. A hint displayed about the necessary reload when running `datacats tweak --add-redis` would be a great failsafe.

**Comment by deniszgonjanin** [comment]
:+1:

**Pull request merged**